diff --git a/feeds_youtube.module b/feeds_youtube.module
index 1f6dc69..50ed016 100755
--- a/feeds_youtube.module
+++ b/feeds_youtube.module
@@ -93,25 +93,16 @@ function feeds_youtube_libraries_info() {
     'name' => 'Google API',
     'vendor url' => 'http://developers.google.com',
     'download url' => 'https://github.com/google/google-api-php-client',
-    'version' => '1.1.3',
+    'version' => '2.2.2',
     'files' => array(
       'php' => array(
-        'src/Google/autoload.php',
-      ),
-    ),
-    'versions' => array(
-      '1.1.3' => array(
-        'files' => array(
-          'php' => array(
-            'src/Google/autoload.php',
-          ),
-        ),
+        '../../../../../vendor/autoload.php',
       ),
     ),
     'version arguments' => array(
       'file' => 'src/Google/Client.php',
       'pattern' => '@LIBVER\s+([0-9\.]+)@',
-      'lines' => 29,
+      'lines' => 42,
       'cols' => 9,
     ),
   );
diff --git a/plugins/FeedsYoutubeParser.inc b/plugins/FeedsYoutubeParser.inc
index d0d7bc2..3d4a49e 100755
--- a/plugins/FeedsYoutubeParser.inc
+++ b/plugins/FeedsYoutubeParser.inc
@@ -23,8 +23,12 @@ class FeedsYoutubeParser extends FeedsParser {
     $libraries = libraries_load('google-api-php-client');
 
     $url = $source->config['FeedsHTTPFetcher']['source'];
+    $parsed_url = parse_url($url);
+    $url_path = ltrim($parsed_url["path"], "/");
+    $path_parts = explode("/", $url_path);
+    $api_type = $path_parts[2];
 
-    $result = new FeedsParserResult();  
+    $result = new FeedsParserResult();
 
     $client = feeds_youtube_googleapi_youtube_client_factory();
     $youtube = new Google_Service_YouTube($client);
@@ -36,7 +40,15 @@ class FeedsYoutubeParser extends FeedsParser {
       $video_results = array();
 
       foreach ($search_response['items'] as $search_result) {
-        array_push($video_results, $search_result['id']['videoId']);
+        switch ($api_type) {
+          case 'search':
+            $videoId = $search_result['id']['videoId'];
+            break;
+          case 'playlistItems':
+            $videoId = $search_result['snippet']['resourceId']['videoId'];
+            break;
+        }
+        array_push($video_results, $videoId);
       }
 
       $video_ids = join(',', $video_results);
@@ -132,7 +144,7 @@ class FeedsYoutubeParser extends FeedsParser {
         'description' => t('Duration of the video in number of seconds.'),
       ),
     );
-    
+
     $private = array(
       'embedded_player' => array(
         'name' => t('Embed Player'),
@@ -151,7 +163,7 @@ class FeedsYoutubeParser extends FeedsParser {
         'name' => t('Number of dislikes'),
       ),
     );
-    if (isset($this->config['authenticated_on_channel']) && 
+    if (isset($this->config['authenticated_on_channel']) &&
       $this->config['authenticated_on_channel'] != 0) {
       $mapping += $private;
     }
@@ -215,8 +227,8 @@ class FeedsYoutubeParser extends FeedsParser {
         'published_datetime' => date('Y-m-d H:i:s', strtotime($published)),
         'published_timestamp' => strtotime($published),
       );
-      
-      if (isset($this->config['authenticated_on_channel']) && 
+
+      if (isset($this->config['authenticated_on_channel']) &&
         $this->config['authenticated_on_channel']) {
 
         $viewCount = $video['statistics']['viewCount'];
@@ -240,7 +252,7 @@ class FeedsYoutubeParser extends FeedsParser {
     return $result;
   }
 
-  
+
   /**
    * Define default configuration.
    */
@@ -249,7 +261,7 @@ class FeedsYoutubeParser extends FeedsParser {
       'authenticated_on_channel' => 0,
     );
   }
-  
+
   /**
    * Build configuration form.
    */
