diff --git a/js/viddler-player.js b/viddler-player.js
new file mode 100644
index 0000000..4db8f69
--- /dev/null
+++ b/js/viddler-player.js
@@ -0,0 +1,25 @@
+;(function ($, Drupal) {
+
+  /**
+   * Render viddler player.
+   * @type {{}}
+   */
+  Drupal.behaviors.viddler = {
+    attach: function (context, settings) {
+      var viddlerId = settings.viddler.viddlerId;
+
+      if (typeof viddlerId !== 'undefined') {
+        Drupal.settings.viddler.exist = [];
+
+        // Add Viddler player object in global Drupal settings.
+        Drupal.settings.viddler.exist['viddler-' + viddlerId] = new ViddlerEmbed({
+          videoId: viddlerId,
+          target: '#viddler-player-' + viddlerId
+        });
+
+        // Delete settings for preventing of rendering viddler player twice and duplicates of videos.
+        delete settings.viddler.viddlerId;
+      }
+    }
+  };
+})(jQuery, Drupal);

